src/app/decorations/qgsdecorationimage.cpp

```diff
@@ -0,0 +1,279 @@
+/***************************************************************************
+  qgsdecorationimage.cpp
+  --------------------------------------
+  Date                 : August 2019
+  Copyright            : (C) 2019 by Mathieu Pellerin
+  Email                : nirvn dot asia at gmail dot com
+ ***************************************************************************
+ *                                                                         *
+ *   This program is free software; you can redistribute it and/or modify  *
+ *   it under the terms of the GNU General Public License as published by  *
+ *   the Free Software Foundation; either version 2 of the License, or     *
+ *   (at your option) any later version.                                   *
+ *                                                                         *
+ ***************************************************************************/
+
+#include "qgsdecorationimage.h"
+
+#include "qgsdecorationimagedialog.h"
+
+#include "qgisapp.h"
+#include "qgscoordinatetransform.h"
+#include "qgsexception.h"
+#include "qgsimagecache.h"
+#include "qgslogger.h"
+#include "qgsmaplayer.h"
+#include "qgsproject.h"
+#include "qgssymbollayerutils.h"
+#include "qgssvgcache.h"
+
+#include <QPainter>
+#include <QMenu>
+#include <QDir>
+#include <QFile>
+#include <QSvgRenderer>
+
+#include <cmath>
+#include <cassert>
+
+
+QgsDecorationImage::QgsDecorationImage( QObject *parent )
+  : QgsDecorationItem( parent )
+{
+  mPlacement = BottomLeft;
+  mMarginUnit = QgsUnitTypes::RenderMillimeters;
+
+  setName( "Image" );
+  projectRead();
+}
+
+void QgsDecorationImage::projectRead()
+{
+  QgsDecorationItem::projectRead();
+  mColor = QgsSymbolLayerUtils::decodeColor( QgsProject::instance()->readEntry( mNameConfig, QStringLiteral( "/Color" ), QStringLiteral( "#000000" ) ) );
+  mOutlineColor = QgsSymbolLayerUtils::decodeColor( QgsProject::instance()->readEntry( mNameConfig, QStringLiteral( "/OutlineColor" ), QStringLiteral( "#FFFFFF" ) ) );
+  mSize = QgsProject::instance()->readDoubleEntry( mNameConfig, QStringLiteral( "/Size" ), 16.0 );
+  setImagePath( QgsProject::instance()->readEntry( mNameConfig, QStringLiteral( "/ImagePath" ), QString() ) );
+  mMarginHorizontal = QgsProject::instance()->readNumEntry( mNameConfig, QStringLiteral( "/MarginH" ), 0 );
+  mMarginVertical = QgsProject::instance()->readNumEntry( mNameConfig, QStringLiteral( "/MarginV" ), 0 );
+}
+
+void QgsDecorationImage::saveToProject()
+{
+  QgsDecorationItem::saveToProject();
+  QgsProject::instance()->writeEntry( mNameConfig, QStringLiteral( "/Color" ), QgsSymbolLayerUtils::encodeColor( mColor ) );
+  QgsProject::instance()->writeEntry( mNameConfig, QStringLiteral( "/OutlineColor" ), QgsSymbolLayerUtils::encodeColor( mOutlineColor ) );
+  QgsProject::instance()->writeEntry( mNameConfig, QStringLiteral( "/Size" ), mSize );
+  QgsProject::instance()->writeEntry( mNameConfig, QStringLiteral( "/ImagePath" ), mImagePath );
```

Comment:
Should path go through the project's path resolver here, so that it respects the project's absolute/relative storage setting? (I think yes)